decidim-core/app/controllers/concerns/decidim/has_stored_path.rb

```diff
@@ -0,0 +1,40 @@
+# frozen_string_literal: true
+
+module Decidim
+  # Shared behaviour for signed_in users that require the latest TOS accepted
+  module HasStoredPath
+    extend ActiveSupport::Concern
+
+    included do
+      # Saves the location before loading each page so we can return to the
+      # right page.
+      before_action :store_current_location
+    end
+
+    # Stores the url where the user will be redirected after login.
+    #
+    # Uses the `redirect_url` param or the current url if there's no param.
+    # In Devise controllers we only store the URL if it's from the params, we don't
+    # want to overwrite the stored URL for a Devise one.
+    def store_current_location
+      return if skip_store_location?
+
+      value = redirect_url || request.url
+      store_location_for(:user, value)
+    end
+
+    def skip_store_location?
+      # Skip if Devise already handles the redirection
+      return true if devise_controller? && redirect_url.blank?
+      # Skip for all non-HTML requests"
+      return true unless request.format.html?
+      # Skip if a signed in user requests the TOS page without having agreed to
+      # the TOS. Most of the times this is because of a redirect to the TOS
+      # page (in which case the desired location is somewhere else after the
+      # TOS is agreed).
+      return true if current_user && !current_user.tos_accepted? && (tos_path.start_with?(request.path) || request.path.starts_with?(tos_path))
```

Comment:
💬 I think you can use the following comparison instead of dealing with `start_with?`. What do you think?

```suggestion
      return true if current_user && !current_user.tos_accepted? && URI(tos_path).path == request.path
```

Reply:
I am adding it like ` URI(request.path).path == URI(tos_path).path` , the main issue is that one of those, is acually comming like : `/proposal?locale=blah` ... and makes it harder to compare :D

Reply:
mm, but `request.path` doesn't include the querystring (`fullpath` does it), there is any other possible difference?

Reply:
I mean, I think you don't need to create a `URI` with `request.path` as it is already 'clean' of modifiers, right?

Reply:
You're right. I have tested and all good now.